### ISSUE TITLE
Regional indicator pairs after a double-width character break cursor position when status is updated.

### ISSUE THREAD

**Issue by lilydjwg** [tmux/tmux#3998]
### Issue description

Paste this string "一🇸🇬" (libc wcswidth=4) into tmux and the cursor position is too left. (It was some other syndrome some commits earlier.)

Minimal tmux.conf:

```
set -g status-interval 1
set -g status-right "%S"
```

(The cursor logically is at the end of the string.)
![图片](https://github.com/tmux/tmux/assets/440661/bc767186-a6d8-4831-bbf7-4d35fb1715bd)

Try to edit the string (not in `cat` but e.g. in bash / python / zsh), more weird behaviors will follow (with which I'm completely lost and can't keep editing).

### Required information

Please provide the following information:

* tmux version (`tmux -V`).

tmux 3.4. 3aa20f6e758a8f96841010d374ed24505d7d89ef is bad, 62f657845ef6229b2f1a78fbceb8f3b2a736227e is good.
 
* Platform (`uname -sp`).

Linux unknown

* $TERM inside and outside of tmux (`echo $TERM`).
 
tmux-256color & xterm-256color.

* Logs from tmux (`tmux kill-server; tmux -vv new`).

[tmux-server-397117.log](https://github.com/user-attachments/files/15750543/tmux-server-397117.log)
[tmux-client-397114.log](https://github.com/user-attachments/files/15750544/tmux-client-397114.log)
[tmux-out-397117.log](https://github.com/user-attachments/files/15750545/tmux-out-397117.log)

**Comment by nicm** [comment]
What terminal is this on? Are you sure it supports VCS?

What does this show inside and outside tmux:
~~~~
printf '|\344\270\200\360\237\207\270\360\237\207\254|\n'
~~~~
Please send me the server log from doing this: `tmux -f/dev/null -vv new` then `printf '|\344\270\200\360\237\207\270\360\237\207\254|\n'` then `exit`.

Are you sure the editing with `cat` doesn't happen outside tmux too?

**Comment by nicm** [comment]
This is U+04E00, U+1F1F8 and U+1F1EC. What does this show in the terminal outside tmux?
~~~~
printf 'X\344\270\200\360\237\207\270\360\237\207\254X';perl -e "print 'x'x(${COLUMNS}-4)";echo
~~~~

**Comment by lilydjwg** [comment]
I'm using GNOME terminal. What is VCS?

Outside:

![图片](https://github.com/user-attachments/assets/2df94f9b-5902-4ffb-8432-53eedb84e6a7)
![图片](https://github.com/user-attachments/assets/42297990-0563-4c1a-933a-b61a931284cf)

Inside:
![图片](https://github.com/user-attachments/assets/48656f9d-b3b5-4869-a259-a19e9be8b1a4)

It looks correctly initially. Only the cursor position would be wrong if there were one.

Logs:

[tmux-out-9416.log](https://github.com/user-attachments/files/17265637/tmux-out-9416.log)
[tmux-server-9416.log](https://github.com/user-attachments/files/17265638/tmux-server-9416.log)
[tmux-client-9414.log](https://github.com/user-attachments/files/17265639/tmux-client-9414.log)

I'm not editing with `cat`, but with python which uses readline. Also weird things happen with zsh & bash inside tmux, but not outside.

This is a screencast. I used a barely-configured bash so it shouldn't be too different. I pasted a string, moved to the beginning with Ctrl-A and then the end with Ctrl-E several times, and then pressed Backspace several times to delete the text.

https://github.com/user-attachments/assets/8d38e026-301a-4e19-8aad-9f63bd024bc0

If I paste the text and delete it again and again, it can go all the way to the left border.

My tmux has updated to version 3.5, and it no longer has the flickering issue as with 3.4, but cursor position is still wrong.

**Comment by lilydjwg** [comment]
I try to use Vim to view the recent "ycombinnator" phishing mail, which contains a lot of U+3164 HANGUL FILLER characters. It causes Vim inside tmux to struggle too. With `less` those characters does not occupy space and cause display issues (less colorful than in Vim for obvious reasons) when scrolling back and forth.

**Comment by nicm** [comment]
Does this still happen with `master`? There were some changes IIRC...

**Comment by lilydjwg** [comment]
I just tried `master`, but no visible change since 3.5.

**Comment by nicm** [comment]
Thanks, I'll take another look.

**Comment by nicm** [comment]
Do you see the same problem if you try another terminal? For example in WezTerm I do not see the `X` overlapping the flag (outside tmux):
<img width="971" height="175" alt="Image" src="https://github.com/user-attachments/assets/cb4bdf8d-f9f2-4246-817e-0bd5f9cb7e8f" />

**Comment by nicm** [comment]
If you change the `variation-selector-always-wide` option in `.tmux.conf`, does it make any difference in GNOME terminal?
~~~~
set -g variation-selector-always-wide off
~~~~

**Comment by lilydjwg** [comment]
I also do not see the overlapping X in wezterm, but I don't think overlapping is a problem: wezterm renders regional indicators as flags so they are narrower, leaving space for the following character. But the cursor position is still confusing:

https://github.com/user-attachments/assets/904986b9-dd7c-4098-9ca9-2958ee3885e5

This is wezterm. What I did is to paste "一🇸🇬" into `zsh` and `cat`, waited for the cursor to jump, and then deleted them one by one with Backspace. The cursor jumps back automatically after about one second (when the status updates?), and the cursor is at the wrong place making editing very hard.

`variation-selector-always-wide` didn't make any difference.

**Comment by nicm** [comment]
Oh I see what you mean, WezTerm is drawing the flag properly as 🇸🇬 and GNOME Terminal is drawing it as SG where both S and G are width 2, but even though they are width 2 and the characters overlap, it is still only taking up 2 cells for the flag character.

So the problem is not the drawing but the cursor positioning...

**Comment by nicm** [comment]
Editing 一🇸🇬 in either `bash` or `cat` seems to work for me, but if I press `C-l` in `bash` or use `C-b r` in tmux it seems to get the cursor in a muddle.

**Comment by nicm** [comment]
Also if I paste it then immediately press the right arrow key the text moves to the right.

**Comment by nicm** [comment]
OK so the problem seems to be:

🇸🇬 is made up of two regional indicators U+1F1F8 (🇸) and U+1F1EC (🇬). The terminal is treating these individually as width 1:

<img width="1081" height="88" alt="Image" src="https://github.com/user-attachments/assets/5e287687-8a91-451f-becb-8eda6818838b" />

But tmux has them set as width 2. You can see this by doing the same thing inside tmux then checking the line end in copy mode:

<img width="764" height="127" alt="Image" src="https://github.com/user-attachments/assets/d900ccd3-a5b9-4afc-a169-12f372cbf892" />

So, when you send 🇸🇬, tmux first receives 🇸and writes it, then it receives 🇬, so it moves back two places to where it thinks the 🇸was, and tries to add the 🇬. But because it should only have moved back one place, it ends up combining the characters incorrectly.

However, that is not the whole story, because even forcing both 🇸 and 🇬 to width 1, it fixes the `printf` above, but it doesn't make 一🇸🇬 work correctly. The next problem is that tmux thinks🇸can be combined as well, and it assumes that because 一 is width 2, it can combine them.

So to fix this we need two changes:

- Change regional indicators alone (not combined with another one) to width 1.

- Only allow them to combine on to other regional indicators, not on to any width 2 character.

Please try this: [regional_indicator.diff.txt](https://github.com/user-attachments/files/23266095/regional_indicator.diff.txt)

**Comment by lilydjwg** [comment]
Your patch works, thanks!

Can you also do something with U+3164 HANGUL FILLER? I've uploaded a html file that doesn't draw correctly in Vim or less.

[b.html.gz](https://github.com/user-attachments/files/23281599/b.html.gz)

**Comment by nicm** [comment]
Great, I have applied it to OpenBSD now - it will be in GitHub later.

I'll take a look at U+3164 next week.

**Comment by nicm** [comment]
@phi4grv do you happen to know what the story is with U+3164? Looks like is an deprecated version of U+115F and U+1160, should it be treated the same?

**Comment by nicm** [comment]
Looks like maybe we can just ignore it? Does this [x.diff.txt](https://github.com/user-attachments/files/23298260/x.diff.txt) fix it?

Note my previous changes have not yet made it to GitHub, hopefully will be there today.

**Comment by lilydjwg** [comment]
I just tried in different programs, but unfortunately things don't align up:

| program | cells occupied |
|---|---|
| Vim | 2 |
| GNOME Terminal | 2 |
| Python (`unicodedata` module) | 2 |
| Alacritty | 0 |
| less | 0 |
| tmux (with x.diff patch) | 0 |
| glibc (`wcwidth`) | -1 |

But at least with the x.diff patch, the display looks consistent while scrolling.

**Comment by nicm** [comment]
I don't think you are going to really get it to work correctly if programs and terminals don't agree on what to do with it. It does seem from a quick look that it should be ignored but I don't write Korean nor know much about Unicode so I don't know what is right here.

**Comment by nicm** [comment]
If the patch makes it better for you I can apply it and we can see if anyone else complains. Nobody else has asked about U+3164 up to now so it doesn't seem like it is that common anymore?

**Comment by lilydjwg** [comment]
Yes, the patch makes it better, please apply it.

**Comment by nicm** [comment]
OK, I have applied this to OpenBSD now, it will be in GitHub eventually (hopefully today). Let me know if you see any other problems. Thanks!
